distributed/scheduler.py

```diff
@@ -7495,16 +7495,28 @@ def decide_worker(
     of bytes sent between workers.  This is determined by calling the
     *objective* function.
     """
+    N_RANDOM_WORKERS: Py_ssize_t = 20
+
     ws: WorkerState = None
     wws: WorkerState
     dts: TaskState
     deps: set = ts._dependencies
+    random_workers_set: set = (
+        valid_workers if valid_workers is not None else all_workers
+    )
     candidates: set
     assert all([dts._who_has for dts in deps])
     if ts._actor:
         candidates = set(all_workers)
     else:
-        candidates = {wws for dts in deps for wws in dts._who_has}
+        if len(random_workers_set) <= N_RANDOM_WORKERS:
+            candidates = random_workers_set
+        else:
+            candidates = {wws for dts in deps for wws in dts._who_has}
```

Comment:
I'm not sure I understand this.  I would have expected something like the following:

```python
if ts._actor:
    ...
else:
    candidates = {wws for dts in deps for wws in dts._who_has}
    candidates |= random_sample(valid_workers or all_workers)
```

Also, as a warning, any time we start getting nested if/elses I'll probably become slightly more critical :)

Reply:
This is an optimization for when there are less than 20 workers, meaning our random sample of workers would actually contain all workers. If we're going to be adding all workers into the candidates set anyway, then candidates will end up containing all workers at the end, so we can skip building candidates from `_who_has`.